Emby.Server.Implementations/Library/Validators/CollectionPostScanTask.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using MediaBrowser.Controller.Collections;
+using MediaBrowser.Controller.Entities;
+using MediaBrowser.Controller.Entities.Movies;
+using MediaBrowser.Controller.Library;
+using MediaBrowser.Model.Querying;
+using Jellyfin.Data.Enums;
+using Microsoft.Extensions.Logging;
+
+namespace Emby.Server.Implementations.Library.Validators
+{
+    /// <summary>
+    /// Class CollectionPostScanTask.
+    /// </summary>
+    public class CollectionPostScanTask : ILibraryPostScanTask
+    {
+        /// <summary>
+        /// The _library manager.
+        /// </summary>
+        private readonly ILibraryManager _libraryManager;
+
+        /// <summary>
+        /// The collection manager.
+        /// </summary>
+        private readonly ICollectionManager _collectionManager;
+
+        /// <summary>
+        /// The logger.
+        /// </summary>
+        private readonly ILogger<CollectionPostScanTask> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CollectionPostScanTask" /> class.
+        /// </summary>
+        /// <param name="libraryManager">The library manager.</param>
+        /// <param name="collectionManager">The collection manager.</param>
+        /// <param name="logger">The logger.</param>
+        public CollectionPostScanTask(
+            ILibraryManager libraryManager,
+            ILogger<CollectionPostScanTask> logger,
+            ICollectionManager collectionManager)
+        {
+            _libraryManager = libraryManager;
+            _collectionManager = collectionManager;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Runs the specified progress.
+        /// </summary>
+        /// <param name="progress">The progress.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>Task.</returns>
+        public async Task Run(IProgress<double> progress, CancellationToken cancellationToken)
+        {
+
+            var movies = _libraryManager.GetItemList(new InternalItemsQuery
+            {
+                IncludeItemTypes = new[] { nameof(Movie) },
+                IsVirtualItem = false,
+                OrderBy = new List<ValueTuple<string, SortOrder>>
+                {
+                new ValueTuple<string, SortOrder>(ItemSortBy.SortName, SortOrder.Ascending)
+                },
+                Recursive = true
+            });
+
+            var boxSets = _libraryManager.GetItemList(new InternalItemsQuery
+            {
+                IncludeItemTypes = new[] { nameof(BoxSet) },
+                CollapseBoxSetItems = false,
+                Recursive = true
+            });
+
+            var numComplete = 0;
+            var count = movies.Count;
+
+            var collectionNameMoviesMap = new Dictionary<string, List<Movie>>();
```

Comment:
You don't seem to be using the `Movie` class for much, so it should probably just be a `List<Guid>`. It will save you a `.Select` later.

Reply:
To implement configureable automiclly create collection, I think `move.ContainingFolderPath` will be needed next. So I'll fix this later.

Reply:
You should be skipping movies that are contained in a library with the option disabled, so `List<Guid>` still makes sense

Reply:
I'm thinking add a minimal variable to library options, when movie amount in same collection is less than this minimal variable, then it won't create or add it to collection.

So in the next foreach map loop, I will need `movie.ContainingFolderPath` to get library option and check whether should add it. It can't be done in first loop, beacause at that time, I don't know how many movies are in same collection.

Reply:
You shouldn't be using `ContainingFolderPath`.

How will you handle two libraries with different collection requirements?

Reply:
Sorry, I'm not follow. What do you mean two libraries with different collection? I think there no other way to find out the movie belong to which library, except using `ContainingFolderPath`.

Reply:
You can have as many Movie libraries as you want, but if the configuration is `MinimumNumberOfMoviesPerCollection`, how do you want to handle if library X is 2 and library Y is 5?

Maybe we'd save ourselves a headache by just forcing it to 2 and then make it a toggle instead (enabled/disabled).

>  I think there no other way to find out the movie belong to which library

We don't have to. LibraryManager.GetLibraryOptions is enough to get the setting.

Reply:
> You can have as many Movie libraries as you want, but if the configuration is MinimumNumberOfMoviesPerCollection, how do you want to handle if library X is 2 and library Y is 5?

Do you mean that one movie in two libraries, and library X is 2 and library Y is 5? That's a problem, but I think we can choose the smallest one. But you are right, the easiesy way is forcing it to 2. We can implement the simple one first.

> We don't have to. LibraryManager.GetLibraryOptions is enough to get the setting.

Thank you, I totally ignore `LibraryManager.GetLibraryOptions`, beacause I saw variable such as `CollectionFolders` in it function, I thought it was only can be used throught collection...

Reply:
If this is complete, the `Movie` object isn't being used to build the collections, so you can just store the ID, as cvium suggested previously

Reply:
Sorry, forgot that. Fixed this in lastest commit.